File: 0417_pacificAtlanticWaterFlow/main.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        vector<vector<int>> res;
        vector<vector<bool>> toPacific(heights.size(),vector<bool>(heights[0].size(), false));
        vector<vector<bool>> toAtlantic(heights.size(),vector<bool>(heights[0].size(), false));

        for(int row = 0; row < heights.size(); ++row)
        {
            dfs(heights, toPacific, row, 0, INT_MIN);
            dfs(heights, toAtlantic, row, heights[row].size()-1, INT_MIN);
        }

        for(int col = 0; col < heights[0].size(); ++col)
        {
            dfs(heights, toPacific, 0, col, INT_MIN);
            dfs(heights, toAtlantic, heights.size()-1, col, INT_MIN);
        }

        for(int row = 0; row < heights.size(); ++row)
        {
            for(int col = 0; col < heights[row].size(); ++col)
            {
                if(toPacific[row][col] && toAtlantic[row][col])
                    res.push_back(vector<int>{row, col});
            }
        }
        return res;
    }

private:
    void dfs(vector<vector<int>>& height, vector<vector<bool>>& isVisited, int row, int col, int prevHeight)
    {
        const int direction[4][2] = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
        if(row < 0 || row >= height.size() || col < 0 || col >= height[0].size() || isVisited[row][col] || prevHeight > height[row][col])
            return;
        
        isVisited[row][col] = true;
        for(int i = 0; i < 4; ++i)
        {
            dfs(height, isVisited, row + direction[i][0], col + direction[i][1], height[row][col]);
        }
    }
};

File: 0417_pacificAtlanticWaterFlow/main.py
from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights:
            return []

        pacific = set()
        atalantic = set()

        def dfs(row, col, land):
            land.add((row, col))
            for i, j in [(row+1, col), (row, col+1), (row-1, col), (row, col-1)]:
                if (0 <= i < len(heights) and 0 <= j < len(heights[0]) and heights[i][j] >= heights[row][col] and (i, j) not in land):
                    dfs(i, j, land)
                    
        for i in range(len(heights)):
            dfs(i, 0, pacific)
            dfs(i, len(heights[0])-1, atalantic)

        for j in range(len(heights[0])):
            dfs(0, j, pacific)
            dfs(len(heights)-1, j, atalantic)

        return list(pacific & atalantic)
                